traverse/visitor.go

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Package traverse provides a visitor framework for analyzing Substrait plans
+with multiple read-only operations in a single tree traversal.
+
+## Key Features:
+- Interface-based visitor pattern
+- Single tree walk for multiple analyses
+- Sequential execution
+- Read-only operations - NO MUTATIONS
+- Zero allocations per node
+
+## IMPORTANT: Read-Only Analysis Only
+This interface is designed for read-only analysis, and assumes the tree does not change
+as it is being walked. Any modification of the tree could disrupt the walking of the
+tree causing nodes to be missed, visited out of order, or cause panics.
```

Comment:
Some modifications would be reasonably safe with this interface.  Here are a few possibilities: https://github.com/voltrondata/spark-substrait-gateway/tree/main/src/transforms